tools/azsdk-cli/Azure.Sdk.Tools.Cli/SampleGeneration/Languages/TypeScriptTypechecker.cs

```diff
@@ -0,0 +1,371 @@
+
+			// Extract packages from various import patterns
+			foreach (var regex in new[] { ImportRegex, RequireRegex, DynamicImportRegex })
+			{
+				foreach (Match match in regex.Matches(code))
+				{
+					var importPath = match.Groups[1].Value;
+
+					// Skip relative imports
+					if (importPath.StartsWith(".") || importPath.StartsWith("/"))
+					{
+						continue;
+					}
+
+					var packageName = GetPackageRoot(importPath);
+					if (!excludedPackages.Contains(packageName))
+					{
+						packages.Add(packageName);
+					}
+				}
+			}
+
+			return packages.ToList();
+		}
+
+		/// <summary>
+		/// Extracts the root package name from an import path.
+		/// Examples: '@scope/pkg/sub/path' → '@scope/pkg', 'pkg/sub/path' → 'pkg', 'fs' → 'fs'
+		/// </summary>
+		private static string GetPackageRoot(string importPath)
+		{
+			if (importPath.StartsWith("@"))
+			{
+				var parts = importPath.Split('/', 3);
+				return parts.Length >= 2 ? $"{parts[0]}/{parts[1]}" : importPath;
+			}
+			return importPath.Split('/')[0];
+		}
+
+		private async Task EnsureContainerAsync(CancellationToken ct)
+		{
+			if (persistentContainer != null)
+			{
+				// Check if container is still running
+				var isRunning = await dockerService.IsContainerRunningAsync(persistentContainer.Name, ct);
+				if (isRunning)
+				{
+					return;
+				}
+
+				// Container exists but is stopped, try to start it
+				logger.LogInformation("Restarting stopped TypeScript verification container: {containerName}", persistentContainer.Name);
+				var restartResult = await dockerService.StartContainerAsync(persistentContainer.Name, ct);
+				if (restartResult.ExitCode == 0)
+				{
+					return;
+				}
+
+				// Failed to start, remove it and create a new one
+				logger.LogWarning("Failed to restart container, removing and creating a new one");
+				await dockerService.RemoveContainerAsync(persistentContainer.Name, force: true, ct);
+				persistentContainer = null;
+			}
+
+			logger.LogInformation("Creating TypeScript verification container");
+
+			// Create a new container
+			var containerName = $"typescript-typecheck-{Guid.NewGuid():N}";
+			var createResult = await dockerService.CreateContainerAsync(
+					DockerImage,
+					containerName,
+					ct: ct);
+
+			if (createResult.ExitCode != 0)
+			{
+				throw new InvalidOperationException($"Failed to create TypeScript container: {createResult.Output}");
+			}
+
+			// Start the container
+			var startResult = await dockerService.StartContainerAsync(containerName, ct);
+			if (startResult.ExitCode != 0)
+			{
+				throw new InvalidOperationException($"Failed to start TypeScript container: {startResult.Output}");
+			}
+
+			persistentContainer = new ContainerInfo(containerName, createResult.Output.Trim());
+			logger.LogInformation("TypeScript verification container created: {containerName}", containerName);
+		}
+
+		private async Task WriteFileToContainerAsync(string containerPath, string content, CancellationToken ct)
```

Comment:
Could we move this API to the docker service? seems like a fairly important functionality to be able to write content directly to a specific location in the container.

Reply:
I looked into it a bit and recalled that there is such a thing already - check out `dockerService.CopyToContainerAsync`. I think the confusion is in the name of the function, as this function is specific to copying the sample into the container (first writing it to a tmp directory on the host, then using `dockerService.CopyToContainerAsync` to copy it into the container).

I renamed the function for more clarity